Proposal: "Concise magic file names" (#1101)
apple/swift-evolution: proposals/NNNN-magic-file.md

# Concise magic file names

* Proposal: [SE-NNNN](NNNN-magic-file.md)
* Authors: [Brent Royal-Gordon](https://github.com/brentdax), [Dave DeLong](https://github.com/davedelong)
* Review Manager: TBD
* Status: **Awaiting review**
* Implementation: [apple/swift#25656](https://github.com/apple/swift/pull/25656); merged behind `-enable-experimental-concise-pound-file`

## Introduction

Today, `#file` evaluates to a string literal containing the full path to the current source file. We propose to instead have it evaluate to a human-readable string containing the filename and module name, while preserving the existing behavior in a new `#filePath` expression.

Swift-evolution thread: [Concise Magic File Names](https://forums.swift.org/t/concise-magic-file-names/31297), [We need `#fileName`](https://forums.swift.org/t/we-need-filename/19781)

## Motivation

In Swift today, the magic identifier `#file` evaluates to a string literal containing the full path (that is, the path passed to `swiftc`) to the current file. It's a nice way to trace the location of logic occurring in a Swift process, but its use of a full path has a lot of drawbacks:

> * It can inadvently reveal private or sensitive information. The full path to a source file may contain a developer's username, hints about the configuration of a build farm, proprietary versions or identifiers, or the Sailor Scout you named an external disk after. Developers probably don't know that this information is embedded in their binaries and may not want it to be there. And most uses of `#file` are in default arguments, which makes this information capture invisible at the use site. The information leaks here are quite serious; if other languages hadn't already normalized this, I doubt we would think that `#file` was an accceptable design.

Reviewer: `inadvently` -> `inadvertently`
`accceptable` -> `acceptable`

```suggestion
* It can inadvertently reveal private or sensitive information. The full path to a source file may contain a developer's username, hints about the configuration of a build farm, proprietary versions or identifiers, or the Sailor Scout you named an external disk after. Developers probably don't know that this information is embedded in their binaries and may not want it to be there. And most uses of `#file` are in default arguments, which makes this information capture invisible at the use site. The information leaks here are quite serious; if other languages hadn't already normalized this, I doubt we would think that `#file` was an acceptable design.
```

Meanwhile, the benefits you might expect from using a full path aren't really there:

* The paths are not portable between machines, so you can't use it to automatically match an error to a line of source code unless your machine has the same directory layout as the machine that built the binary.

* The paths are not guaranteed to be absolute—XCBuild and SwiftPM happen to use absolute paths, but other build systems like Bazel don't—so you can't be sure that you can resolve paths to a file on disk.
  
* The paths are not even guaranteed to be unique within a process because the same file could be included in several different modules, or different projects could be located at the same paths on different machines or at different times.

For `#file`'s most important use case, it seems like the current string computed by `#file` is suboptimal and a better default would improve Swift in several ways.

### Other uses of `#file`

> While `#file` is primarily intended for developer-facing errors, in practice it is also used in other ways. In particular, Swift tests sometimes use `#file` to compute paths to fixtures relative to the source file that uses them. This has historically been necessary in SwiftPM because it did not support resources, but [SE-0271][] has now added a much better solution to this problem than `#file`-based path computation.

Reviewer: Use relative URL?

```suggestion
While `#file` is primarily intended for developer-facing errors, in practice it is also used in other ways. In particular, Swift tests sometimes use `#file` to compute paths to fixtures relative to the source file that uses them. This has historically been necessary in SwiftPM because it did not support resources, but [SE-0271](0271-package-manager-resources.md) has now added a much better solution to this problem than `#file`-based path computation.
```

>   [SE-0271]: https://github.com/apple/swift-evolution/blob/master/proposals/0271-package-manager-resources.md

Reviewer: Remove reference? Use relative URL above?

2. `<StaticString typealias> = #file`: A couple of projects use a custom typealias for `StaticString`. We manually confirmed that these projects generally pass those values to something like an `XCTAssert` function which should show a concise string.

3. `String = #file`, but with `#line` on the same line: We take this to indicate that a file and line are being captured for display together.

4. `#file` interpolated into a string: Since `#file` is usually an absolute path that wouldn't need to be concatenated, we take these to be formatting for display.

We also identified two patterns that we conservatively assume represent eventual I/O or other machine-oriented processing (although in many cases they may represent path computation for display):

1. `String = #file`, no `#line` nearby: We assume this will be passed to an API like `URL.init(fileURLWithPath:)` which will then be used to further manipulate the path or perform I/O.

> 2. `#file` used in a parethensized list, but not an interpolation: We assume this is being passed to an API like `URL.init(fileURLWithPath:)`.

Reviewer: `parethensized` -> `parenthesized`

```suggestion
2. `#file` used in a parenthesized list, but not an interpolation: We assume this is being passed to an API like `URL.init(fileURLWithPath:)`.
```
